homeassistant/components/smartthings/audio.py

```diff
@@ -0,0 +1,242 @@
+
+    pcm: bytes
+    created: float
+    expires: float
+
+
+class SmartThingsAudioManager(HomeAssistantView):
+    """Manage PCM proxy URLs for SmartThings audio notifications."""
+
+    url = "/api/smartthings/audio/{token}"
+    name = "api:smartthings:audio"
+    requires_auth = False
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize the manager."""
+        self.hass = hass
+        self._entries: OrderedDict[str, _AudioEntry] = OrderedDict()
+        self._lock = asyncio.Lock()
+
+    async def async_prepare_notification(self, source_url: str) -> str:
+        """Generate an externally accessible PCM URL for SmartThings."""
+        pcm, duration, truncated = await self._transcode_to_pcm(source_url)
+        if not pcm:
+            raise SmartThingsAudioError("Converted audio is empty")
+
+        if truncated:
+            _LOGGER.warning(
+                "SmartThings audio notification truncated to %s seconds (output length %.1fs); longer sources may be cut off",
+                FFMPEG_MAX_DURATION_SECONDS,
+                duration,
+            )
+        elif duration > WARNING_DURATION_SECONDS:
+            _LOGGER.warning(
+                "SmartThings audio notification is %.1fs; playback over %s seconds may be cut off",
+                duration,
+                WARNING_DURATION_SECONDS,
+            )
+
+        token = secrets.token_urlsafe(16)
+        now = dt_util.utcnow().timestamp()
+        entry = _AudioEntry(
+            pcm=pcm,
+            created=now,
+            expires=now + ENTRY_TTL.total_seconds(),
+        )
+
+        async with self._lock:
+            self._cleanup(now)
+            self._entries[token] = entry
+            self._entries.move_to_end(token)
+            while len(self._entries) > MAX_STORED_ENTRIES:
+                dropped_token, _ = self._entries.popitem(last=False)
+                _LOGGER.debug(
+                    "Dropped expired SmartThings audio token %s", dropped_token
+                )
+
+        path = f"/api/smartthings/audio/{token}{PCM_EXTENSION}"
+        try:
+            base_url = get_url(
+                self.hass,
+                allow_internal=True,
+                allow_external=True,
+                allow_cloud=True,
+                prefer_external=True,
+                prefer_cloud=True,
+            )
+        except NoURLAvailableError as err:
+            async with self._lock:
+                self._entries.pop(token, None)
+            raise SmartThingsAudioError(
+                "SmartThings audio notifications require an accessible Home Assistant URL"
+            ) from err
+
+        return f"{base_url}{path}"
+
+    async def get(self, request: web.Request, token: str) -> web.StreamResponse:
+        """Serve a PCM audio response."""
+        token = token.removesuffix(PCM_EXTENSION)
+
+        async with self._lock:
+            now = dt_util.utcnow().timestamp()
+            self._cleanup(now)
```

Comment:
I don't like the cleanup in the get request.
we can get the token and check if it's valid. See previous comment